### ISSUE TITLE
mercurius references `Window` type that is not always available

### ISSUE THREAD

**Issue by runeh** [mercurius-js/mercurius#419]
Mercurius uses types from `Window`. That type is not available in typescript
unless the `lib` setting is set to `DOM` or is omitted.

See testcase here: https://github.com/runeh/mercurius-window-bug-repro

Steps to reproduce:

- `npm install`
- `npm run build`

Expected:

Compiles

Actual:

Typescript fails with

```
node_modules/mercurius/index.d.ts:545:33 - error TS2304: Cannot find name 'Window'.

545   playgroundHeaders?: ((window: Window) => object) | object;
                                    ~~~~~~
```

**Comment by mcollina** [comment]
Would you like to send a Pull Request to address this issue?

Removing it is totally ok.

**Comment by mrazauskas** [comment]
For me it seems that all is correct in the code.

The reported issue is related with typings of `playgroundHeaders` option. Documentation of the option: `If it is a function, it is serialised, injected in the generated HTML and invoked with the 'window' object as the argument.` It will not work without `Window`.

Mercurius is Node library. It includes two browser side IDEs: GraphiQL and GraphQL Playground. These are the reason why `"lib": "DOM"` is necessary in `tsconfig.json`. See #367.

**Comment by mrazauskas** [comment]
Apologies. I concentrated on Mercurius only to explain why there is no need to change anything. This did not answer the original question.

To make the example repo build, add `"skipLibCheck": true` in your `tsconfig.json`:

```json
{
  "compilerOptions": {
    "lib": ["ES2019"],
    "skipLibCheck": true
  }
}
```

Without this setting `tsc` will check all *.d.ts files including `node_modules/mercurius/index.d.ts` as we can see in the error. These checks make compile time slower too.

If you prefer to keep library checks, you have to add `"lib": "DOM"` in your `tsconfig.json`. As I was explaining previously, Mercurius typings depend DOM definitions.

**Comment by runeh** [comment]
It should not be required to use `skipLibCheck: true` in order to use mercurius. I explicitly turn that setting OFF in all my projects because I want typescript to complain about potential errors.

Furthermore, mercurius does not run in a browser, thus enabling the DOM lib does not make sense. To quote  the typescript documentation https://www.typescriptlang.org/tsconfig#lib :

> Your program doesn’t run in a browser, so you don’t want the "dom" type definitions

Most importantly, having the DOM lib enabled means typescript can no longer prevent runtime errors if I use browser only objects. With the DOM lib enabled, typescript allows this code to compile:

```typescript
window.alert();
```

However, it will crash at runtime.

**Comment by mcollina** [comment]
https://github.com/mercurius-js/mercurius/pull/420 should fix this.

**Comment by mrazauskas** [comment]
I agree with all what is said. At the same time we should keep in mind that Mercurius was written in JS and typings were added later. I hope you will agree that in this case best practise is to have typings which reflect the code.

Perhaps it makes sense to rework the code instead of changing uncomfortable parts of typings into `any` to make `tsc` happy?

#306 fits here too. Unfortunately implementation of IDEs is clumsy and needs to be reworked. I was writing a motivation which suggest to remove IDEs and offer them as plugins (`altair-fastify-plugin` is perfect example, also this suits well with Fastify). That’s braking change. Might be people will get angry. Better to think about it slowly.

Here is quick experiment. I removed all Mercurius typings related with IDEs: `graphiql`, `ide`, `playgroundSettings`, `playgroundHeaders`. This is how typings would look if IDEs got removed. No browser served code. Trying to build the example repo throws an errors, I have to delete this line (of course, `playground` is not included):

```diff
app.register(mercurius, {
  schema,
  resolvers,

-  graphiql: 'playground'
});
```

And it builds. No need for `"lib": "DOM"` or `"skipLibCheck": "true"`.

To try out plugin idea let’s add `altair-fastify-plugin` and let’s build. All works.

This issue extends the motivation which I am writing. Really good that it came in and revealed one more issue related with current IDEs implementation. But I do not see reasons to alter typings right now.
